csharp/extractor/Semmle.Extraction.CSharp.Standalone/Runtime.cs

```diff
@@ -3,38 +3,107 @@
         private static string ExecutingRuntime => RuntimeEnvironment.GetRuntimeDirectory();
 
-        /// <summary>
-        /// Locates .NET Core Runtimes.
-        /// </summary>
-        private static IEnumerable<string> CoreRuntimes
+        public Runtime(IDotNet dotNet) => this.dotNet = dotNet;
+
+        internal sealed class RuntimeVersion : IComparable<RuntimeVersion>
         {
-            get
+            private readonly string dir;
+            private Version Version { get; }
+            private Version? PreviewVersion { get; } = null;
+            private bool IsPreview => PreviewVersion is not null;
+            public string FullPath
             {
-                var dotnetPath = FileUtils.FindProgramOnPath(Win32.IsWindows() ? "dotnet.exe" : "dotnet");
-                var dotnetDirs = dotnetPath is not null
-                    ? new[] { dotnetPath }
-                    : new[] { "/usr/share/dotnet", @"C:\Program Files\dotnet" };
-                var coreDirs = dotnetDirs.Select(d => Path.Combine(d, "shared", "Microsoft.NETCore.App"));
+                get
+                {
+                    var preview = IsPreview ? $"-preview.{PreviewVersion}" : "";
+                    var version = Version + preview;
+                    return Path.Combine(dir, version);
+                }
+            }
 
-                var dir = coreDirs.FirstOrDefault(Directory.Exists);
-                if (dir is not null)
+            private static Version MakeVersion(string version)
+            {
+                var versionParts = version.Split('.');
+                return new Version(int.Parse(versionParts[0]), int.Parse(versionParts[1]), int.Parse(versionParts[2]));
+            }
+
+            public RuntimeVersion(string dir, string version, string previewVersion)
+            {
+                this.dir = dir;
+                Version = MakeVersion(version);
+                if (!string.IsNullOrEmpty(previewVersion))
                 {
-                    return Directory.EnumerateDirectories(dir).OrderByDescending(Path.GetFileName);
+                    PreviewVersion = MakeVersion(previewVersion);
                 }
+            }
 
-                return Enumerable.Empty<string>();
+            public int CompareTo(RuntimeVersion? other)
+            {
+                var c = Version.CompareTo(other?.Version);
+                if (c == 0 && IsPreview)
+                {
+                    return other!.IsPreview ? PreviewVersion!.CompareTo(other!.PreviewVersion) : -1;
+                }
+                return c;
             }
+
+            public override bool Equals(object? obj) =>
+                obj is not null && obj is RuntimeVersion other && other.FullPath == FullPath;
+
+            public override int GetHashCode() => FullPath.GetHashCode();
+
+            public override string ToString() => FullPath;
+        }
+
+        [GeneratedRegex(@"^(\S+)\s(\d+\.\d+\.\d+)(-preview\.(\d+\.\d+\.\d+))?\s\[(\S+)\]$")]
```

Comment:
I don't think we should hardcode `preview`. I've seen different prerelease variants. See for example https://github.com/dotnet/ef6/commit/dcc62bf98b4af45cbff52ec1168d93eaebab3b7e#diff-8df3cd354bc584349d04ad5675b33c042d8b99b741b8b95af394c55e0f5001bf.

I've seen `alpha`, `rc`, `preview`. I don't yet know how these are ordered. Maybe simply alphabetically. `alpha`, `beta`, `preview`, `rc` seem to be in order.

Reply:
Ah - great - I will also try to incorporate that.